File: 3775.reverse-words-with-same-vowel-count.cpp
// @leet start
#include <string>
#include <utility>
using namespace std;
class Solution {
private:
    bool is_vowel(char c) {
        return c == 'a' or c == 'e' or c == 'i' or c == 'o' or c == 'u';
    }

    void reverse(string& s, int start, int end) {
        for (int i = 0; i < (end - start + 1) / 2; ++i) {
            swap(s[start + i], s[end - i]);
        }
    }

public:
    string reverseWords(string s) {
        int num_vowels = 0;

        int i = 0;
        for (; i < s.size() and s[i] != ' '; ++i) {
            if (is_vowel(s[i])) { ++num_vowels; }
        }

        int start = i + 1, end = start;
        while (end < s.size()) {
            int curr_num_vowels = 0;
            for (end = start; end < s.size() and s[end] != ' '; ++end) {
                if (is_vowel(s[end])) { ++curr_num_vowels; }
            }
            if (curr_num_vowels == num_vowels) { reverse(s, start, end - 1); }

            start = end + 1;
        }

        return s;
    }
};
// @leet end

File: 3775.reverse-words-with-same-vowel-count.rs
// @leet start
impl Solution {
    pub fn reverse_words(s: String) -> String {
        let mut bytes = s.into_bytes();
        let n = bytes.len();

        let is_vowel = |b: u8| -> bool { matches!(b, b'a' | b'e' | b'i' | b'o' | b'u') };

        let mut first_end = 0;
        let mut target_vowel_num = 0;
        while first_end < n && bytes[first_end] != b' ' {
            if is_vowel(bytes[first_end]) {
                target_vowel_num += 1;
            }

            first_end += 1;
        }

        let mut left = first_end + 1;
        while left < n {
            let mut right = left;
            let mut curr_vowel_num = 0;

            while right < n && bytes[right] != b' ' {
                if is_vowel(bytes[right]) {
                    curr_vowel_num += 1;
                }

                right += 1;
            }

            if curr_vowel_num == target_vowel_num {
                bytes[left..right].reverse();
            }

            left = right + 1;
        }

        return String::from_utf8(bytes).unwrap();
    }
}
// @leet end
